README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# knnstdisagg

An R package to perform space and time disaggregation of streamflow
using a K-nearest neighbor (knn) approach.

[![Project Status: Active – The project has reached a stable, usable
state and is being actively
developed.](https://www.repostatus.org/badges/latest/active.svg)](https://www.repostatus.org/#active)
[![R build
status](https://github.com/rabutler-usbr/knnstdisagg/workflows/R-CMD-check/badge.svg)](https://github.com/rabutler-usbr/knnstdisagg/actions)
[![Codecov test
coverage](https://codecov.io/gh/rabutler-usbr/knnstdisagg/branch/master/graph/badge.svg)](https://codecov.io/gh/rabutler-usbr/knnstdisagg?branch=master)

## Installation

Currently only available on GitHub

``` r
# install.packages("remotes")
remotes::install_github("rabutler-usbr/knnstdisagg")
```

## Example Usage

One application of the KNN space-time disaggregation methodology is to
take paleo reconstructed data at Lees Ferry in the Colorado River Basin,
and disaggregate those annual flows at one location to monthly flows at
29 locations. The following steps through how to use the knnstdisagg
package to do so.

As the space-time disaggregation method relies on monthly pattern data,
we need monthly pattern data, and we will use the
[CoRiverNF](https://github.com/BoulderCodeHub/CoRiverNF) package for
those data.

``` r
remotes::install_github("BoulderCodeHub/CoRiverNF")
```

### Setup the data

The space-time disaggregation disaggregates an annual value (`ann_flow`)
to monthly data by matching `ann_flow` to an annual index value
(`ann_index_flow`). Then, the monthly pattern and spatial pattern
(`mon_flow`) from the selected annual index year is used to disaggregate
the data.

In this example, we will be disaggregating the Meko et al. (2007) paleo
reconstructed data (`meko`), which is provided as an example dataset in
this package:

``` r
library(knnstdisagg)
library(CoRiverNF)
#> Loading required package: xts
#> Loading required package: zoo
#> 
#> Attaching package: 'zoo'
#> The following objects are masked from 'package:base':
#> 
#>     as.Date, as.Date.numeric
#> CoRiverNF package currently includes 1906-2018 natural flows, released January 10, 2020.

head(meko)
#>      water_year lees_ferry
#> [1,]        762   16360000
#> [2,]        763   13190000
#> [3,]        764   12140000
#> [4,]        765   12690000
#> [5,]        766   16660000
#> [6,]        767   14380000
```

`meko` is already formatted correctly for use in this package: one
column of years and one column of annual data.

The annual index flow and monthly values are from the CoRiverNF package.
We will match the `meko` data to the historical water year data at Lees
Ferry. Additionally, for now, the index data needs to be a two column
matrix and not an xts object. For the monthly data, we need it to be
full water years, so need to remove the last three months of data so it
stops at the end of the last water year.

``` r
# setup annual data
annual_index <- CoRiverNF::wyAnnTot$LeesFerry
yrs <- as.numeric(format(index(wyAnnTot$LeesFerry), "%Y"))
annual_index <- as.matrix(annual_index)
annual_index <- cbind(yrs, annual_index)

# setup monthly data
last_month <- paste0("/", max(yrs), "-09")
monthly_data <- CoRiverNF::monthlyInt[last_month]
```

*Note, this example uses a named xts object for the monthly data, which
means the results are also a named xts object. An unnamed matrix will
also work.*

### Space-time disaggregation

The space-time disaggregation is performed by `knn_space_time_disagg()`.
We have already setup the data necesary for `ann_flow`,
`ann_index_flow`, and `mon_flow`. Because this is water year data, we
will set the `start_month` to 10 as the water year starts in October. We
will only disaggregate the data once, so there is only one “simulation”.
In previous work, we have found that we want to scale the Upper Basin
nodes based on the volume at Lees Ferry, but in the Lower Basin, we will
not scale their values, i.e., we will select the monthly data directly
for the selected index year. The Upper Basin sites are sites 1-20.
Finally, we will use the default weighting scheme from Nowak et al. to
select the nearest neighbor.

``` r
disagg <- knn_space_time_disagg(
  ann_flow = meko,
  ann_index_flow = annual_index,
  mon_flow = monthly_data,
  start_month = 10,
  nsim = 1,
  scale_sites = 1:20,
  k_weights = knn_params_default(nrow(annual_index))
)
#> Sites 21, 22, 23, 24, 25, 26, 27, 28, 29
#> will be selected directly from the index years, i.e., not scaled.
```

The results are now in `disagg`, and we can get the output using
`knnst_get_disagg_data()`:

``` r
head(knnst_get_disagg_data(disagg)[,5:10]) # only look at a few sites
#>         Crystal GrandJunction CiscoDolores CiscoColorado Fontenelle
#> 761-10 5034.272      84700.62    39670.062      6224.374   65747.59
#> 761-11 4631.530      36982.77    19135.267     14840.026   45308.45
#> 761-12 1610.967      34120.28     9108.005     10993.843   39267.32
#> 762-01 2114.394      27275.68     8234.055      3563.258   37253.61
#> 762-02 2617.821      34888.51    11017.000      5985.749   32219.34
#> 762-03 4832.901      33166.79    32489.177      9785.618   61418.12
#>        GreenRiverWY
#> 761-10    12585.680
#> 761-11     9263.060
#> 761-12     4027.417
#> 762-01     2919.878
#> 762-02     4128.103
#> 762-03    21445.998
```

If needed, the output can be saved to disk using `write_knnst()`. This
saves the disaggregated data for every simulation as well as the
selected index years.

### QA/QC

#### Base Statistics

The knnstdisagg package also includes plotting functionality to assist
with QA/QC. Plots of monthly statistics (mean, max, min, variance, lag-1
correlation, and skew), annual statistics (same as monthly), annual cdf,
and a monthly cdf for each month can be created using `plot()`. Each
call to plot works for one site. A `bin_size` must be specified; this is
the moving window that all statistics on the disaggregated data are
computed accross. Looking at Glenwood Springs monthly statistics:

``` r
p <- plot(
  disagg, 
  site = "GlenwoodSprings", 
  base_units = "acre-feet", 
  which = 14, 
  show = TRUE,
  bin_size = 50
)
```

<img src="man/figures/README-unnamed-chunk-9-1.png" width="100%" />

*If an unnamed matrix was used for input, then the sites are accessed by
`"S1"`, `"S2"`, etc. during plotting.*

We can also look at the annual cdf for Bluff, or the May cdf for
Maybell:

``` r
p <- plot(
  disagg, 
  site = "Bluff", 
  base_units = "acre-feet", 
  which = 13, 
  show = TRUE,
  bin_size = 50
)
```

<img src="man/figures/README-unnamed-chunk-11-1.png" width="100%" />

``` r
p <- plot(
  disagg, 
  site = "Maybell", 
  base_units = "acre-feet", 
  which = 5, 
  show = TRUE,
  bin_size = 50
)
```

<img src="man/figures/README-unnamed-chunk-13-1.png" width="100%" />

All plots for a given site can be created at once using `which = 1:15`,
and the suite of plots can be saved using `save_knnstplot()`. For
example, all plots for the data at Cameo could be saved to a pdf using:

``` r
p <- plot(
  disagg, 
  site = "Cameo", 
  base_units = "acre-feet", 
  which = 1:15, 
  show = FALSE,
  bin_size = 50
)
save_knnstplot(p, "Cameo.pdf", width = 8, height = 6)
```

#### Spatial Correlation

Another statistic to check is the spatial correlation between sites.
This is computed using `knnst_spatial_cor()`. For each specified site,
the correlation with all other sites is computed, and then it can be
easily plotted. To get the correlation from Cameo and Hoover to all
other
sites:

``` r
sp_cor <- knnst_spatial_cor(disagg, sites = c("Cameo", "Hoover"), bin_size = 50)
plot(sp_cor)
```

<img src="man/figures/README-unnamed-chunk-15-1.png" width="100%" />

#### Temporal Correlation

A final statistic to compare is the monthly cross correlation. This is
computed using `knnst_temporal_cor()` and is only computed for one site
at a time. To compute the monthly cross correlation at Greendale:

``` r
tmp_cor <- knnst_temporal_cor(disagg, site = "Greendale", bin_size = 50)
plot(tmp_cor)
```

<img src="man/figures/README-unnamed-chunk-16-1.png" width="100%" />

### Recreating Disaggregation

**Will be added to vignette**

## Acknowledgements

This package implements the methods developed by Nowak et al. (2010):

Nowak, K., J. Prairie, B. Rajagopalan, and U. Lall (2010), A
nonparametric stochastic approach for multisite disaggregation of annual
to daily streamflow, Water Resour. Res., 46, W08529,
[doi:10.1029/2009WR008530](https://agupubs.onlinelibrary.wiley.com/doi/abs/10.1029/2009WR008530).

It also uses the Meko et al. (2007) paleo reconstructed Lees Ferry
natural flow as example data in the package and above:

Meko, D.M., Woodhouse, C.A., Baisan, C.A., Knight, T., Lukas, J.J.,
Hughes, M.K., and Salzer, M.W. 2007. Medieval Drought in the Upper
Colorado River Basin. Geophysical Research Letters 34, L10705.

## Disclaimer

This software is in the public domain because it contains materials that
originally came from the U.S. Bureau of Reclamation, an agency of the
United States Department of Interior.

Although this code has been used by Reclamation, no warranty, expressed
or implied, is made by Reclamation or the U.S. Government as to the
accuracy and functioning of the program and related program material nor
shall the fact of distribution constitute any such warranty, and no
responsibility is assumed by Reclamation in connection therewith.

This software is provided “AS IS.”


NEWS.md
# knnstdisagg 0.2.1

**Released September 7, 2023**

## Minor non API breaking changes

* `knnst_spatial_cor()`, `knnst_temporal_cor()` (#34), and `plot.knnst()` now work when there is more than one simulation. (#34)
* Small syntax/function updates for newer versions of ggplot2 and tidyr

## Bug fixes

* Fixed labeling of PDFs (they were previously labeled CDFs) (#35)
* Fixed formatting of README generated by `write_knnst()` (#36)

# knnstdisagg 0.2.0

**Released April 16, 2020**

## API Breaking Changes

* `plot.knnst()` was changed in multiple ways, several which break any previous calls to it. 
    * The default value for `site` was removed as it does not make sense to guess site names when they will typically be unique.
    * Stats for the disaggregated data are now computed on a user specified `bin_size`. `bin_size` is a new argument to `plot.knnst()` with no default. (#32)
    * The use of `...` was modified so that it is now only used for color, size, and shape. Default color, size, and shape were added for instances when these are not user specified.

## Additions and Enhancements

* Added spatial and temporal correlation functions and plots. (#29, #30)
  * `knnst_spatial_cor()` computes spatial correlation between sites.
  * `knnst_temporal_cor()` computes temporal (monthly) correlation at one site.
  * `plot.knnst_spcor()` and `plot.knnst_tmpcor()` plot the results of `knn_spatial_cor()` or `knn_temporal_cor()`
* Added `random_seed` argument to `knn_space_time_disagg()` and `knn_get_index_year()` to assist with reproducibility. (#28)
* Changed how unnamed `mon_flow` input to `knn_space_time_disagg()` is treated. Now the columns (sites) are renamed in `knn_space_time_disagg()`, instead of in `as.data.frame.knnst()`. This ensures the site names always exist for any subsequent code, rather than having to convert to a data.frame to get the site names. (#33)
* `write_knnst()` now saves a README.txt file (#25) and includes the yyyy-mm row names in the saved csv files that include the monthly disaggregated data.
* `ex_disagg` is now an exported data object. It contains an example `knnst` object for use in examples and tests.

## Bug Fixes

* Fixed bug related to how lag-1 correlation was computed for the historical/pattern data. This only affected January values if the start of the year was a month other than January. (#27)


# knnstdisagg 0.1.0

**Released April 7, 2020**

First release of package. Includes annual to monthly disaggregation and plots for QA/QC + diagnostics.


R/plot.knnst.R
#' Plot resulting statistics for knn space-time disaggregation
#'
#' `plot.knnst()` implements the [plot()] method for `knnst` objects, and relies
#' on ggplot2. It plots the key statistics the knn space-time disaggregation
#' method should be preserving.
#'
#' If there are more  than one simulations, the stats are computed for each
#' simulation across all years. If there is only one simulation, then the user
#' must specify the `bin_size` that is used to compute the stats across
#' moving windows of this specified size.
#'
#' `which` controls the plots that are created. There are both monthly and
#' annual plots and two main plot types: a pdf of the flows across all of the
#' different disaggregations, and a panel showing the mean, max, min, variance,
#' lag-1 correlation, and skew across all of the different disaggregations. The
#' numbers correspond to:
#'
#' - `1:12`: the different monthly pdfs.
#' - `13`: annual pdf
#' - `14`: monthly statistics panel
#' - `15`: annual statistics panel
#'
#' `...` is passed to [geom_point()] and [geom_line()] to overwrite defaults for
#' `size`, `shape`, and `color`.
#'
#' For the monthly statistics panel, months are ordered based on the
#' `start_month` provided in the original disaggregation, i.e., specified by the
#' user in [knn_space_time_disagg()].
#'
#' @param x An object inheriting from class `knnst`
#'
#' @param site The site to plot. Site name as a character.
#'
#' @param bin_size Number of years for each bin when there is only one
#'   simulation.
#'
#' @param base_units The (input) units of the flow that was disaggregated. These
#'   units will be shown as the y-axis label of the plot.
#'
#' @param which The subset of plots to create; specify a subset of the number
#'   `1:15`. See 'Details' for the different plots.
#'
#' @param show Boolean that determines if diagnostic plots are shown or only
#'   returned. If `TRUE` and in interactive mode, will be able to enter through
#'   each of the 1-15 plots.
#'
#' @param ... Arguments to be passed to subsequent methods.
#'
#' @return A `knnstplot` object
#'
#' @export
plot.knnst <- function(x, site, bin_size = NULL, base_units = NULL,
                       which = c(13, 14, 15), show = FALSE, ...)
{
  #TODO: update to handle multiple simulations
  # assert_that(
  #   knnst_nsim(x) == 1,
  #   msg = "plot.knnst() is currently only setup to work with one simulation."
  # )

  assert_that(
    length(site) == 1 && is.character(site),
    msg = "In `plot.knnst()`, `site` should be a character with length of 1."
  )

  assert_that(
    length(which) > 0 && is.numeric(which) && all(which %in% 1:15),
    msg = "In `plot.knnst()`, `which` should be numeric values in 1:15"
  )

  assert_that(
    is.logical(show) && length(show) == 1,
    msg = "In `plot.knnst()`, `show` should be a logical scalar."
  )

  nsim <- knnst_nsim(x)
  nyrs <- nrow(x$disagg_sims[[1]]$disagg_flow) / 12

  if (nsim == 1) {
    assert_that(is.numeric(bin_size) && length(bin_size) == 1)
  }

  # if there is more than 1 simulation, then set bin_size to full length of data
  if (nsim > 1) {
    message('More than 1 simulation exists, so stats computed for all years in each simulation.')
    bin_size <- nyrs
  }

  assert_that(bin_size <= nyrs)

  x_df <- as.data.frame(x)
  all_cols <- names(x_df)

  assert_that(
    site %in% names(x_df),
    msg = "In `plot.knnst()`, `site` should be a valid site name."
  )

  # set flags for needing monthly/annual data
  are_mon <- are_ann <- FALSE
  if (any(which %in% c(1:12, 14)))
    are_mon <- TRUE

  if (any(which %in% c(13, 15)))
    are_ann <- TRUE

  # simulation stats ----------------
  if (are_mon)
    x_plot_data <- get_mon_plot_stats(
      x_df,
      site = site,
      start_month = x$start_month,
      bin_size = bin_size,
      yr = "year"
    )

  if (are_ann) {
    x_ann_plot_data <- get_ann_plot_stats(x_df, site, x$start_month, bin_size)
    x_ann_sim_data <- x_df %>%
      dplyr::group_by_at(c("year", "simulation")) %>%
      dplyr::summarise_at(site, sum)
  }
  # get historical stats -------------------
  # get historical data as a data frame, organize, and compute same stats
  # as on the simulated data
  nhist_yrs <- nrow(x$mon_flow) / 12
  x_mon <- get_pattern_flow_data_df(x, site)

  if (are_mon) {
    x_mon_stats <- get_mon_plot_stats(
      x_mon,
      site = site,
      start_month = x$start_month,
      bin_size = nhist_yrs,
      yr = "agg_year"
    )
  }
  # annual historical data
  if (are_ann) {
    x_ann <- get_historical_annual(x_mon, site, x$start_month)
    x_ann_stats <- get_ann_plot_stats(x_mon, site, x$start_month, nhist_yrs)
  }

  gg1 <- gg2 <- gg3 <- gg4 <- NULL
  # monthly plots ----------------
  if (14 %in% which)
    gg1 <- create_mon_bxp(x_plot_data, x_mon_stats, site, x$start_month,
                          base_units, bin_size, nsim, ...)

  if (any(1:12 %in% which))
    gg2 <- create_mon_pdf(x_df, x_mon, bin_size, site, base_units, which,
                          nsim, ...)

  # annual plots --------------------
  if (15 %in% which)
    gg3 <- create_ann_bxp(x_ann_plot_data, x_ann_stats, site, base_units,
                          bin_size, nsim, ...)

  if (13 %in% which)
    gg4 <- create_ann_pdf(x_ann_sim_data, x_ann, bin_size, site, base_units,
                          nsim, ...)

  gg_out <- c(
    list(
      "monthly-stats" = gg1,
      "annual-stats" = gg3,
      "annual-pdf" = gg4
    ),
    gg2
  )
  gg_out <- structure(gg_out, class = "knnstplot")

  if (show && interactive())
    print(gg_out)

  invisible(gg_out)
}

# yr is which year variable to keep.
get_mon_plot_stats <- function(x_df, site, start_month, bin_size, yr = "year")
{
  keep_cols <- c("ym", yr, "month", site, "simulation")
  vars_group <- c("month", "simulation")

  # if yr is year, then convert it to agg_year
  # convert year to agg_year
  if (yr == "year") {
    x_df[["year"]] <- x_df[["ym"]]
    x_df <- dplyr::mutate_at(
      x_df,
      "year",
      list(~get_agg_year(., start_month))
    )
  }

  x_df %>%
    # subset to site
    dplyr::select_at(keep_cols) %>%
    dplyr::group_by_at("simulation") %>%
    dplyr::arrange_at("ym") %>%
    # compute stats for all months
    get_plot_stats(
      var_mutate = site, vars_group = vars_group, bin_size = bin_size, yr = yr
    ) %>%
    dplyr::ungroup() %>%
    dplyr::mutate_at(
      "month",
      list(~ factor(month.abb[.], levels = month.abb))
    )
}

get_ann_plot_stats <- function(x_df, site, start_month, bin_size)
{

  x_df[["agg_year"]] <- x_df[["ym"]]

  x_df %>%
    # get the "agg_year" from ym
    dplyr::mutate_at("agg_year", list(~ get_agg_year(., start_month))) %>%
    # sum to annual values
    dplyr::select_at(c("agg_year", site, "simulation")) %>%
    dplyr::group_by_at(c("agg_year", "simulation")) %>%
    dplyr::summarise_at(site, list("ann" = sum)) %>%
    # compute the same stats as monthly
    dplyr::group_by_at("simulation") %>%
    dplyr::arrange_at("agg_year") %>%
    get_plot_stats(var_mutate = "ann", vars_group = "simulation", bin_size, "agg_year")

}

# for a single variable `var_mutate`, compute the mean, variance, max, min,
# lag-1 correlation, and skew
get_plot_stats <- function(x_df, var_mutate, vars_group, bin_size, yr)
{
  var_name_order <- c(
    "mean" = "Mean",
    "stats::var" = "Variance",
    "max" = "Maximum",
    "min" = "Minimum",
    "stats::cor" = "Lag-1 Correlation",
    "skew" = "Skew"
  )

  # it is already arranged by ym, so get the agg_year from the first and last
  # entries to compute number of years
  start_year <- min(x_df[[yr]])
  nbin <- max(x_df[[yr]]) - start_year + 1 - bin_size + 1

  res <- data.frame()

  # TODO: make this more efficient!!!!!
  # pre-allocating an array and then converting to df took 1.75 seconds in a
  # simple test, will staying as a data frame and using bind_rows to 3.08
  # seconds

  for (i in seq_len(nbin)) {
    yr_st <- start_year + i - 1
    yr_end <- yr_st + bin_size - 1

    tmp <- x_df %>%
      dplyr::filter_at(yr, dplyr::all_vars(. %in% yr_st:yr_end)) %>%
      dplyr::mutate_at(var_mutate, list("tmp" = dplyr::lag)) %>%

      dplyr::group_by_at(vars_group) %>%
      # means, standard deviation, max, min, skew, lag-1 correlation
      dplyr::summarise_at(
        var_mutate,
        list(
          ~ mean(.), ~ stats::var(.), ~ max(.), ~ min(.), ~ skew(.),
          ~ stats::cor(., get("tmp"), use = "complete.obs")
        )
      ) %>%
      tidyr::pivot_longer(
        -tidyselect::all_of(vars_group),
        names_to = 'Variable',
        values_to = 'Value'
      ) %>%
      dplyr::mutate_at(
        "Variable",
        list(~ factor(var_name_order[.], levels = var_name_order))
      )

    tmp[["bin"]] <- i

    res <- dplyr::bind_rows(res, tmp)
  }

  res
}

create_ann_bxp <- function(sim_data, hist_data, site, base_units = NULL,
                           bin_size, nsim, ...)
{
  shape <- plot_ops("shape", ...)
  color <- plot_ops("color", ...)
  size <- plot_ops("size", ...)

  gg <- ggplot(sim_data, aes(y = .data[["Value"]])) +
    geom_boxplot() +
    facet_wrap("Variable", ncol = 2, scales = "free_y") +
    geom_point(
      data = hist_data,
      aes(x = 0, y = .data[["Value"]]),
      shape = shape,
      color = color,
      size = size
    ) +
    labs(
      x = NULL,
      title = paste(site, "- Annual Statistics"),
      y = paste("Base units =", base_units),
      caption = caption_text(bin_size, nsim, "points")
    ) +
    theme(axis.ticks.x = element_blank(), axis.text.x = element_blank()) +
    scale_x_continuous(breaks = 0) +
    coord_cartesian( xlim = c(-1, 1))

  gg
}

create_mon_bxp <- function(sim_data, hist_data, site, start_month,
                           base_units = NULL, bin_size, nsim, ...)
{
  shape <- plot_ops("shape", ...)
  color <- plot_ops("color", ...)
  size <- plot_ops("size", ...)

  sim_data[["month"]] <- factor(
    sim_data[["month"]],
    levels = month.abb[full_year(start_month)]
  )

  hist_data[["month"]] <- factor(
    hist_data[["month"]],
    levels = month.abb[full_year(start_month)]
  )

  gg <- ggplot(sim_data, aes(.data[["month"]], .data[["Value"]])) +
    geom_boxplot(aes(group = .data[["month"]])) +
    facet_wrap("Variable", ncol = 2, scales = "free_y") +
    geom_point(
      data = hist_data,
      aes(.data[["month"]], .data[["Value"]]),
      shape = shape,
      size = size,
      color = color
    ) +
    labs(
      x = NULL,
      title = paste0(site, " - Monthly Statistics"),
      y = paste("Base units =", base_units),
      caption = caption_text(bin_size, nsim, "points")
    )

  gg
}

create_ann_pdf <- function(sim_data, hist_data, bin_size, site, base_units,
                           nsim, ...)
{
  pdf_data <- compute_pdf_data(sim_data, hist_data, bin_size, site)

  gg <- pdf_plot(pdf_data, base_units, title = paste(site, "- Annual PDF"),
                 bin_size, nsim, ...)

  gg
}

create_mon_pdf <- function(sim_data, hist_data, bin_size, site, base_units,
                           which, nsim, ...)
{
  # TODO: do we need to re-do this based on Balaji/Ken's code? Using density
  # instead of hist()
  # compute histograms for all simulations
  # 1) call density() initially
  # find which months are supposed to be processed
  proc_mon <- 1:12
  proc_mon <- proc_mon[proc_mon %in% which]

  mon_labels <- month.abb
  names(mon_labels) <- 1:12

  sim_data <- dplyr::arrange_at(sim_data, "ym")
  hist_data <- dplyr::arrange_at(hist_data, "ym")

  gg <- list()

  for (mm in proc_mon) {
    tmp <- compute_pdf_data(
      dplyr::filter_at(sim_data, "month", dplyr::all_vars(. == mm)),
      dplyr::filter_at(hist_data, "month", dplyr::all_vars(. == mm)),
      bin_size,
      site
    )

    # TODO: should boxplot width be computed?
    pname <- paste0(month.abb[mm], "-pdf")
    gg[[pname]] <- pdf_plot(
      tmp,
      base_units,
      title = paste0(site, " - ", month.name[mm], " PDF"),
      bin_size,
      nsim,
      ...
    )
  }

  gg
}

# computes pdf data for one site and one month/year. Assumes sim_data and
# hist_data have already been filtered to correct month/years
compute_pdf_data <- function(sim_data, hist_data, bin_size, site)
{
  # compute the breaks ----
  tmp <- c(hist_data[[site]], sim_data[[site]])
  mon_breaks <- stats::density(tmp, n = 25)$x

  # create all bins --------
  # this is essentially annual data
  all_bins <- get_all_bins(sim_data[[site]], bin_size, monthly = FALSE)

  # call stats::density on each bin -------
  dd <- apply(
    all_bins,
    2,
    stats::density,
    n = 25,
    from = min(mon_breaks),
    to = max(mon_breaks)
  )

  # get all the x, y values from every column and
  sim_data_res <- do.call(rbind, lapply(1:ncol(all_bins), function(i) {
    data.frame(
      "x" = dd[[i]]$x,
      "density" = dd[[i]]$y
    )
  }))

  hist_data_res <- stats::density(
    hist_data[[site]],
    n = 25,
    from = min(mon_breaks),
    to = max(mon_breaks)
  )
  hist_data_res <- data.frame(
    x = hist_data_res$x,
    density = hist_data_res$y
  )

  list(sim_pdf = sim_data_res, hist_pdf = hist_data_res)
}

# assumes x is alreadly order correctly. Given data x, and the bin_size,
# return a matrix of all the data which would have bin_size rows (*12 for
# monthly), and nrow(x) - bin_size + 1 columns
get_all_bins <- function(x, bin_size, monthly = TRUE)
{
  mult <- ifelse(monthly, 12, 1)
  nbins <- length(x) / mult - bin_size + 1

  select_data <- function(start_index, zz, bin_size, monthly) {
    if (monthly) {
      start_index <- start_index * 12 - 11
      zz <- zz[start_index:(bin_size * 12 + start_index - 1)]
    } else {
      # annual data
      zz <- zz[start_index:(bin_size + start_index - 1)]
    }

    zz
  }

  all_bins <- simplify2array(
    lapply(1:nbins, select_data, x, bin_size, monthly)
  )

  all_bins
}

# given a yyyy-mm and start_month, determine its "aggregation" year. ex water
# year or fiscal year.
get_agg_year <- function(x, start_month)
{
  x <- simplify2array(strsplit(x, "-", fixed = TRUE))
  yy <- as.numeric(x[1,])
  mm <- as.numeric(x[2,])

  if (start_month != 1)
    yy[mm %in% start_month:12] <-  yy[mm %in% start_month:12] + 1

  yy
}

get_historical_annual <- function(x, site, start_month)
{
  x[["agg_year"]] <- x[["ym"]]
  x %>%
    dplyr::mutate_at("agg_year", list(~ get_agg_year(., start_month))) %>%
    dplyr::group_by_at(c("agg_year", "simulation")) %>%
    dplyr::summarise_at(site, sum)
}

pdf_plot <- function(pdf_data, base_units, title, bin_size, nsim, ...)
{
  color = plot_ops("color", ...)

  ggplot(pdf_data$sim_pdf, aes(.data[["x"]], .data[["density"]])) +
    geom_boxplot(aes(group = .data[["x"]])) +
    geom_line(
      data = pdf_data$hist_pdf,
      aes(.data[["x"]], .data[["density"]]),
      color = color
    ) +
    labs(
      title = title,
      x = paste0("Flow (", base_units, ")"),
      y = "Probability Density",
      caption = caption_text(bin_size, nsim, "line")
    )
}

caption_text <- function(bin_size, nsim, pl = "point")
{

  if (nsim == 1) {
    paste0("Colored ", pl, " = input/pattern; boxplots = ", bin_size,
           "-year moving window on disaggregated data")
  } else {
    paste0("Colored ", pl, " = input/pattern; boxplots = stats computed across entire period for each simulation.")
  }
}

# checks ... to see if something was specified, otherwise sets it to default
plot_ops <- function(op, ...)
{
  args <- list(...)
  if (exists(op, where = args)) {
    x <- args[[op]]
  } else {
    defaults <- list(
      color = "#51B2FF",
      size = 2,
      shape = 17
    )

    x <- defaults[[op]]
  }

  x
}
